flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/Executing.java

```diff
@@ -184,6 +226,10 @@ void goToFailing(
                 ExecutionGraphHandler executionGraphHandler,
                 OperatorCoordinatorHandler operatorCoordinatorHandler,
                 Throwable failureCause);
+
+        int getAvailableSlots();
+
+        Optional<? extends VertexAssignment> getVertexAssignmentWithAllSlots();
```

Comment:
I think we can simplify this a bit by introducing a `boolean canScaleUp` method.